### PR TITLE
Handle the Drool Rule Validation in RAS to Create ORG Role Assignment…

### PR DESCRIPTION
**Pull request by adewumio-hmcts**
… in lower env

**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/AM-1194

### Change description ###
Handle the Drool Rule Validation in RAS to Create ORG Role Assignment in lower env

Introduce a new attribute in application.yml
Introduce an new attribute in Request.java
Set new value in Request.java after creation of object in ParseRequestService.java with boolean value read from application.yml

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X] No
```

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
     @Test
     void shouldApproveOrgRequestedRoleForTCW_S001() {
         //clientId check not implemented yet
-        //assignmentRequest.getRequest().setClientId("am_org_role_mapping_service");
+        assignmentRequest.getRequest().setClientId("am_org_role_mapping_service");
```
Comment (dharmendra8186) [inline src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java:24]: Write more test case to support positive and negative scenarios with  the help of  byPassOrgDroolRule  flag in request.

**Review by dharmendra8186 — COMMENTED** [review]
Please write integration test cases to support drool rule use cases for this change.

File: src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java (hunk 1 above)
Comment (adewumio-hmcts) [reply src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java:30]: Done

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/ParseRequestService.java [hunk 2]
```diff
@@ -41,11 +42,15 @@
...
 
+    @Value("${org.request.byPassOrgDroolRule}")
+    private boolean byPassOrgDroolRule;
+
     public AssignmentRequest parseRequest(AssignmentRequest assignmentRequest, RequestType requestType)
         throws ParseException {
         long startTime = System.currentTimeMillis();
         logger.info(String.format("parseRequest execution started at %s", startTime));
         Request request = assignmentRequest.getRequest();
+        request.setByPassOrgDroolRule(byPassOrgDroolRule);
```
[the remark this answers is not in the document]
Comment (adewumio-hmcts) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/ParseRequestService.java:53]: Done

File: charts/am-role-assignment-service/values.prod.template.yaml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+java:
+  environment:
```
[the remark this answers is not in the document]
Comment (adewumio-hmcts) [reply charts/am-role-assignment-service/values.prod.template.yaml]: Done

File: src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentCreateAndDeleteIntegrationTest.java [hunk 4]
```diff
@@ -234,11 +240,36 @@ public void shouldDeleteRoleAssignmentsByAssignmentId() throws Exception {
...
+    private AssignmentRequest buildDroolRuleBypassRequest() throws Exception {
+        final AssignmentRequest assignmentRequest =
+            TestDataBuilder.createRoleAssignmentRequest(
+            true,
+            true
+        );
+        assignmentRequest.getRequest().setRequestType(RequestType.CREATE);
+        // assignmentRequest.getRequest().setStatus(CREATE_REQUESTED);
+        assignmentRequest.setRequestedRoles(getRequestedOrgRole());
+        assignmentRequest.getRequestedRoles().stream().forEach(roleAssignment -> {
```
[the remark this answers is not in the document]
Comment (dharmendra8186) [reply src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentCreateAndDeleteIntegrationTest.java:262]: this is unused method. Please remove it.

File: src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentCreateAndDeleteIntegrationTest.java [hunk 5]
```diff
@@ -265,9 +296,14 @@ private Integer getAssignmentRecordsCount() {
 
     @NotNull
     private HttpHeaders getHttpHeaders() {
+        return getHttpHeaders(AUTHORISED_SERVICE);
```
Comment (dharmendra8186) [inline src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentCreateAndDeleteIntegrationTest.java:299]: why you create this method?

File: src/integrationTest/resources/sql/insert_org_assignment_records_to_delete.sql [hunk 6]
```diff
@@ -0,0 +1,19 @@
+DELETE FROM role_assignment_history;
+DELETE FROM role_assignment_request;
+DELETE FROM role_assignment;
+DELETE FROM actor_cache_control;
+
```
Comment (dharmendra8186) [inline src/integrationTest/resources/sql/insert_org_assignment_records_to_delete.sql:5]: Why you create  this file if there is no use ?

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/model/Assignment.java [hunk 7]
```diff
@@ -45,6 +45,10 @@
     private JsonNode notes;
     private List<String> authorisations;
 
+    public void setStatus(Status status) {
```
Comment (dharmendra8186) [inline src/main/java/uk/gov/hmcts/reform/roleassignment/domain/model/Assignment.java:48]: no need to create set method. As , we are using lombok annotation to handle these.

File: src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/DroolJudicialCategoryTest.java [hunk 8]
```diff
@@ -81,6 +81,7 @@ void shouldApprovedRequestedRoleForOrg() {
         assignmentRequest.getRequestedRoles().stream().forEach(roleAssignment -> {
             roleAssignment.setRoleCategory(RoleCategory.JUDICIAL);
             roleAssignment.setRoleType(RoleType.ORGANISATION);
+            roleAssignment.setStatus(Status.CREATE_REQUESTED);
```
Comment (dharmendra8186) [inline src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/DroolJudicialCategoryTest.java:84]: why we need to change here?

File: src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/DroolJudicialCategoryTest.java [hunk 9]
```diff
@@ -108,6 +109,7 @@ void shouldRejectOrgValidation_MissingAttributeJurisdiction() {
             roleAssignment.setRoleType(RoleType.ORGANISATION);
             roleAssignment.setRoleName("judge");
             roleAssignment.setGrantType(STANDARD);
+            roleAssignment.setStatus(Status.CREATE_REQUESTED);
```
Comment (dharmendra8186) [inline src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/DroolJudicialCategoryTest.java:112]: same here..

File: src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java [hunk 10]
```diff
@@ -88,13 +98,15 @@ void shouldRejectOrgRequestedRoleForTCW_WrongClientID_S003() {
...
         assignmentRequest.setRequestedRoles(getRequestedOrgRole());
         assignmentRequest.getRequestedRoles().stream().forEach(roleAssignment -> {
             roleAssignment.setRoleCategory(RoleCategory.JUDICIAL);
             roleAssignment.setRoleType(RoleType.ORGANISATION);
             roleAssignment.setRoleName("tribunal-caseworker");
             roleAssignment.setGrantType(STANDARD);
+            roleAssignment.setStatus(CREATE_REQUESTED);
+            roleAssignment.setRoleType(CASE);
```
Comment (dharmendra8186) [inline src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java:109]: why you are setting roletype= case in staff org test file? Please remove it.

File: src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java [hunk 11]
```diff
@@ -111,13 +123,14 @@ void shouldRejectOrgValidationForTCW_WrongRoleCategory_S004() {
...
         //clientId check not implemented yet
-        //assignmentRequest.getRequest().setClientId("am_org_role_mapping_service");
+        assignmentRequest.getRequest().setClientId("am_org_role_mapping_service");
         assignmentRequest.setRequestedRoles(getRequestedOrgRole());
         assignmentRequest.getRequestedRoles().stream().forEach(roleAssignment -> {
             roleAssignment.setRoleCategory(RoleCategory.STAFF);
             roleAssignment.setRoleType(RoleType.ORGANISATION);
             roleAssignment.setRoleName("tribunal-caseworker");
             roleAssignment.setGrantType(SPECIFIC);
+            roleAssignment.setStatus(CREATE_REQUESTED);
```
Comment (dharmendra8186) [inline src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java:133]: what's need to set status here?

File: src/test/java/uk/gov/hmcts/reform/roleassignment/helper/TestDataBuilder.java [hunk 12]
```diff
@@ -533,7 +533,7 @@ public static ExistingRoleAssignment buildExistingRoleForIAC(String actorId, Str
                                  .actorIdType(ActorIdType.IDAM)
                                  .classification(Classification.PUBLIC)
                                  .readOnly(true)
-                                 .status(CREATE_REQUESTED)
+                                 .status(CREATED)
```
Comment (dharmendra8186) [inline src/test/java/uk/gov/hmcts/reform/roleassignment/helper/TestDataBuilder.java:536]: why you change it? we are checking the status CREATED_REQUESTED in drool rule.

File: test.json [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+      "id": "a67f4278-9e09-4d55-98ed-ac56bb1ad3af",
+      "authenticatedUserId": "6b36bfc6-bb21-11ea-b3de-0242ac130006",
+      "correlationId": "38a90097-434e-47ee-8ea1-9ea2a267f51d",
+      "assignerId": "123e4567-e89b-42d3-a456-556642445678",
+      "requestType": "CREATE",
+      "process": "S-052",
+      "reference": "S-052",
+      "replaceExisting": true,
+      "status": "REJECTED",
+      "created": "2021-02-23T17:19:29.0742Z",
```
Comment (dharmendra8186) [inline test.json:13]: what's the need to create this file if not using anywhere?

**Review by dharmendra8186 — COMMENTED** [review]
Please check code review comments and fix it.

File: src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/DroolPassAssignmentCreateAndDeleteIntegration.java [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    private static final String ASSIGNMENT_ID = "f7edb29d-e421-450c-be66-a10169b04f0a";
+    private static final String ACTOR_ID = "123e4567-e89b-42d3-a456-556642445612";
+    public static final String CREATED = "CREATED";
+    public static final String APPROVED = "APPROVED";
+    public static final String LIVE = "LIVE";
+    public static final String DELETED = "DELETED";
+    public static final String DELETE_APPROVED = "DELETE_APPROVED";
+    private static final String AUTHORISED_SERVICE = "ccd_gw";
+
+    private MockMvc mockMvc;
```
[the remark this answers is not in the document]
Comment (adewumio-hmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/DroolPassAssignmentCreateAndDeleteIntegration.java:70]: It was meant to handle instances where @TestPropertySource(properties = {"org.request.byPassOrgDroolRule=false"})
The other IntegrationTest class already covers the case where org.request.byPassOrgDroolRule=true

File: src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentCreateAndDeleteIntegrationTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (adewumio-hmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentCreateAndDeleteIntegrationTest.java:262]: Removed

File: src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentCreateAndDeleteIntegrationTest.java (hunk 5 above)
Comment (adewumio-hmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentCreateAndDeleteIntegrationTest.java:299]: This is to provide.a means of injecting different clientId values.  ClientId values are been taken from that header value in the securityutil class.

Comment (adewumio-hmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentCreateAndDeleteIntegrationTest.java:299]: It was  used at some point but it's now removed

File: src/integrationTest/resources/sql/insert_org_assignment_records_to_delete.sql (hunk 6 above)
Comment (adewumio-hmcts) [reply src/integrationTest/resources/sql/insert_org_assignment_records_to_delete.sql:5]: Removed

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/model/Assignment.java (hunk 7 above)
Comment (adewumio-hmcts) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/domain/model/Assignment.java:48]: It was introduced to help with debugging, thanks for spotting it.  Removed now

File: src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/DroolJudicialCategoryTest.java (hunk 9 above)
Comment (adewumio-hmcts) [reply src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/DroolJudicialCategoryTest.java:112]: This was added so that the request will meet the drool rule condition as the test case it's trying to achieve

File: src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java (hunk 10 above)
Comment (adewumio-hmcts) [reply src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java:109]: Done

File: src/test/java/uk/gov/hmcts/reform/roleassignment/helper/TestDataBuilder.java (hunk 12 above)
Comment (adewumio-hmcts) [reply src/test/java/uk/gov/hmcts/reform/roleassignment/helper/TestDataBuilder.java:536]: Done

File: src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java (hunk 11 above)
Comment (adewumio-hmcts) [reply src/test/java/uk/gov/hmcts/reform/roleassignment/domain/service/drools/StaffCategoryOrgRoleTest.java:133]: Done

**Review by aashishaggarwal123 — APPROVED** [review]
